src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -75,12 +75,33 @@ Status CoreWorkerDirectTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
             task_spec.GetSchedulingClass(), task_spec.GetDependencyIds(),
             task_spec.IsActorCreationTask() ? task_spec.ActorCreationId()
                                             : ActorID::Nil());
-        auto it = task_queues_.find(scheduling_key);
-        if (it == task_queues_.end()) {
-          it =
-              task_queues_.emplace(scheduling_key, std::deque<TaskSpecification>()).first;
+        auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+        scheduling_key_entry.task_queue.push_back(task_spec);
+        if (scheduling_key_entry.total_tasks_in_flight <
+            scheduling_key_entry.active_workers.size() *
+                max_tasks_in_flight_per_worker_) {
+          // The pipelines to the current workers are not full yet, so we don't need more
+          // workers.
+
+          // Find a worker with a number of tasks in flight that is less than the maximum
+          // value (max_tasks_in_flight_per_worker_) and call OnWorkerIdle to send tasks
+          // to that worker
+          for (auto active_worker_addr : scheduling_key_entry.active_workers) {
+            RAY_CHECK(worker_to_lease_entry_.find(active_worker_addr) !=
+                      worker_to_lease_entry_.end());
+            auto &lease_entry = worker_to_lease_entry_[active_worker_addr];
+            if (lease_entry.tasks_in_flight < max_tasks_in_flight_per_worker_) {
```

Comment:
Suggest wrapping this in a const method of the WorkerLeaseEntry to make it more readable!